File: latte/src/main/java/gg/beemo/latte/broker/BrokerMessage.kt
package gg.beemo.latte.broker

import gg.beemo.latte.broker.rpc.RpcMessageHeaders
import gg.beemo.latte.broker.rpc.RpcRequestMessage
import gg.beemo.latte.broker.rpc.RpcResponseMessage
import gg.beemo.latte.broker.rpc.RpcStatus

open class BrokerMessage<T, H : BrokerMessageHeaders>(
    val topic: String,
    val key: String,
    val value: T,
    val headers: H
) {

    val messageId: String
        get() = headers.messageId

    internal fun <ResponseT> toRpcRequestMessage(
        updateSender: suspend (RpcStatus, ResponseT) -> Unit,
    ): RpcRequestMessage<T, ResponseT, H> {
        return RpcRequestMessage(topic, key, value, headers, updateSender)
    }

    internal fun toRpcResponseMessage(): RpcResponseMessage<T> {
        return RpcResponseMessage(topic, key, value, RpcMessageHeaders(headers))
    }

}

typealias AbstractBrokerMessage<T> = BrokerMessage<T, out BrokerMessageHeaders>
typealias BaseBrokerMessage<T> = BrokerMessage<T, BrokerMessageHeaders>

File: water/src/broker/BrokerMessage.ts
import type { BrokerMessageHeaders, RpcStatus } from "@";
import { RpcMessageHeaders, RpcRequestMessage, RpcResponseMessage } from "@";

export class BrokerMessage<T, H extends BrokerMessageHeaders = BrokerMessageHeaders> {
	public constructor(
		public readonly topic: string,
		public readonly key: string,
		public readonly value: T,
		public readonly headers: H,
	) {}

	public get messageId(): string {
		return this.headers.messageId;
	}

	// Private API
	public toRpcRequestMessage<ResponseT>(
		updateSender: (status: RpcStatus, data: ResponseT) => Promise<void>,
	): RpcRequestMessage<T, ResponseT, H> {
		return new RpcRequestMessage(this.topic, this.key, this.value, this.headers, updateSender);
	}

	// Private API
	public toRpcResponseMessage(): RpcResponseMessage<T> {
		return new RpcResponseMessage(this.topic, this.key, this.value, new RpcMessageHeaders(this.headers));
	}
}
